core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -378,9 +393,42 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://stackoverflow.com/questions/52816049/how-to-convert-cbdescriptor-value-to-string
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        when (updatedDescriptor.UUID.UUIDString) {
+            CBUUIDCharacteristicFormatString -> {
+                return updatedDescriptor.value as NSData
+            }
+            CBUUIDCharacteristicAggregateFormatString,
+            CBUUIDCharacteristicUserDescriptionString,
+            -> {
+                return (updatedDescriptor.value as String)
+                    .encodeToByteArray()
+                    .toNSData()
+            }
+            CBUUIDCharacteristicExtendedPropertiesString,
+            CBUUIDClientCharacteristicConfigurationString,
+            CBUUIDServerCharacteristicConfigurationString,
+            CBUUIDL2CAPPSMCharacteristicString,
+            -> {
+                // https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-54/out/en/host/generic-attribute-profile--gatt-.html
+                // should be two bytes in length
+                val rawValue = (updatedDescriptor.value as NSNumber).intValue
+                return byteArrayOf(
+                    (rawValue.ushr(1) and 0xFF).toByte(),
+                    (rawValue and 0xFF).toByte(),
+                ).toNSData()
+            }
+            else -> {
+                logger.warn { message = "cannot read descriptor for unknown uuid string ${updatedDescriptor.UUID.UUIDString}" }
+                return byteArrayOf().toNSData()
+            }
+        }
```

Comment:
This whole section, despite being very useful, seems out-of-scope of this PR (belongs in a separate PR)?

Additionally, being that these are UUIDs for standard descriptors, it probably makes sense to provide similar functionality for the other platforms as well, right?

Reply:
Yes you are right, this section was added because I've encountered another issue after I've solved the original descriptor discovery problem.
I'll open a separate PR for this.

> Additionally, being that these are UUIDs for standard descriptors, it probably makes sense to provide similar functionality for the other platforms as well, right?

Yes, I need to verify the behavior on other platforms and maybe is possible to provide common code for this.

Reply:
I've just opened #706 for this :)